mne/tests/test_source_tfr.py

```diff
@@ -0,0 +1,225 @@
+    assert_array_equal(stfr.times, stfr.tmin + np.arange(n_times) * stfr.tstep)
+
+    assert_allclose(stfr.times,
+                    [0., 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])
+
+    def attempt_times_mutation(stfr):
+        stfr.times -= 1
+
+    def attempt_assignment(stfr, attr, val):
+        setattr(stfr, attr, val)
+
+    # .times is read-only
+    with pytest.raises(RuntimeError,
+                       match="cannot write to the .times attribute directly"):
+        attempt_times_mutation(stfr)
+    with pytest.raises(RuntimeError,
+                       match="cannot write to the .times attribute directly"):
+        attempt_assignment(stfr, "times", [1])
+
+    # Changing .tmin or .tstep re-computes .times
+    stfr.tmin = 1
+    assert (type(stfr.tmin) == float)
+    assert_allclose(stfr.times,
+                    [1., 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9])
+
+    stfr.tstep = 1
+    assert (type(stfr.tstep) == float)
+    assert_allclose(stfr.times,
+                    [1., 2., 3., 4., 5., 6., 7., 8., 9., 10.])
+
+    # tstep <= 0 is not allowed
+    with pytest.raises(ValueError, match="must be greater than 0"):
+        attempt_assignment(stfr, 'tstep', 0)
+    with pytest.raises(ValueError, match="must be greater than 0"):
+        attempt_assignment(stfr, 'tstep', -1)
+
+    # Changing .data re-computes .times
+    stfr.data = np.random.rand(100, 20, 5)
+    assert_allclose(stfr.times, [1., 2., 3., 4., 5.])
+
+    # .data must match the number of vertices
+    with pytest.raises(ValueError, match="must match the number of vertices"):
+        attempt_assignment(stfr, "data", [[[1]]])
+
+    # .data much match number of dimensions
+    assign = [None, np.arange(100), [np.arange(100)], [[[np.arange(100)]]]]
+    for val in assign:
+        with pytest.raises(ValueError,
+                           match="Data.*? should have.*? dimensions"):
+            attempt_assignment(stfr, "data", val)
+
+    # .shape attribute must also work when ._data is None
+    stfr._kernel = np.zeros((2, 2))
+    stfr._sens_data = np.zeros((2, 5, 3))
+    stfr._data = None
+    assert_equal(stfr.shape, (2, 5, 3))
+
+    # bad size of data
+    stfr = fake_stfr
+    data = stfr.data[:, :, np.newaxis, :]
+    with pytest.raises(ValueError, match='3 dimensions for SourceTFR'):
+        SourceTFR(data, stfr.vertices)
+    stfr = SourceTFR(data[:, :, :, 0], stfr.vertices, 0, 1)
+    assert stfr.data.shape == (data.shape[0], data.shape[1], 1)
+
+
+@requires_h5py
+def test_io_stfr_h5(fake_stfr, fake_kernel_stfr):
+    """Test IO for stfr files using HDF5."""
+    for stfr in [fake_stfr, fake_kernel_stfr]:
+        tempdir = _TempDir()
+        with pytest.raises(ValueError, match="can only be written as HDF5"):
+            stfr.save(op.join(tempdir, 'tmp'), ftype='foo')
+        out_name = op.join(tempdir, 'tempfile')
+        stfr.save(out_name, ftype='h5')
+        stfr.save(out_name, ftype='h5')  # test overwrite
```

Comment:
you should check the round-trip. aka you write, you read and you make sure everything is fine.